src/codegen/compile.ml

```diff
@@ -9864,85 +9758,87 @@ let compile_Nat64_kernel env name op shortcut =
 
 (* helper, expects i64 on stack *)
 let enforce_32_unsigned_bits env =
-  compile_bitand64_const 0xFFFFFFFF00000000L ^^
-  G.i (Test (Wasm.Values.I64 I64Op.Eqz)) ^^
+  compile_bitand_const 0xFFFF_FFFF_0000_0000L ^^
+  compile_test I64Op.Eqz ^^
   else_arithmetic_overflow env
 
 (* helper, expects two identical i64s on stack *)
 let enforce_32_signed_bits env =
-  compile_shl64_const 1L ^^
-  G.i (Binary (Wasm.Values.I64 I64Op.Xor)) ^^
+  compile_shl_const 1L ^^
+  G.i (Binary (Wasm_exts.Values.I64 I64Op.Xor)) ^^
   enforce_32_unsigned_bits env
 
 let compile_Int32_kernel env name op =
```

Comment:
Since Int32 and Nat32 are now always compact, we could probably refactor compile_smallInt_Kernel and compile_small_Nat_Kernel to do the job of compile_Int32_kernel and compile_Nat32_kernel and all the special casing on Int32 and Nat32 in compile_binop . I'd just add a TODO since changing any maths makes me nervous and we might want to keep them apart if we support both 64 and 32 bit in one compile.ml

Reply:
Absolutely, this would make the design cleaner in 64-bit. I added the TODO, and we can realize once we have the single 64-bit backend.